### ISSUE TITLE
Adressing random mask impact on edge relevance (GNNExplainer)

### ISSUE THREAD

**Issue by joaquincabezas** [pyg-team/pytorch_geometric#1985]
## ❓ Questions & Help

Hi everyone,

while looking at the edge_mask values for different runs on the same model/dataset, but with different random seeds, I found some of the edges do show a high variance. For example, using the script at https://github.com/rusty1s/pytorch_geometric/blob/master/examples/gnn_explainer.pym, we explain node 51 (which has 12 edges on its 2-neighborhood). We execute the explanation 9 times and obtain the following values (rows are runs, columns are edges):

1. [0.9103, 0.8305, 0.0631, 0.7547, 0.0736, 0.0740, 0.0717, 0.0749, 0.9186, 0.9592, 0.0490, 0.8682]
2. [0.9269, 0.8785, 0.0714, 0.8732, 0.0684, 0.0692, 0.9394, 0.9641, 0.8821, 0.0670, 0.0529, 0.7739]
3. [0.0364, 0.8383, 0.9277, 0.8577, 0.0637, 0.0709, 0.9372, 0.9428, 0.0404, 0.0677, 0.9278, 0.8978]
4. [0.9082, 0.8983, 0.0693, 0.8961, 0.9313, 0.9313, 0.0750, 0.0590, 0.9126, 0.9238, 0.0656, 0.8617]
5. [0.8800, 0.8025, 0.0618, 0.9010, 0.0729, 0.0656, 0.9312, 0.0741, 0.8824, 0.0659, 0.9246, 0.8525]
6. [0.9053, 0.8527, 0.0651, 0.8179, 0.9317, 0.9319, 0.9311, 0.9330, 0.9170, 0.0592, 0.0678, 0.8915]
7. [0.9167, 0.8516, 0.0682, 0.9054, 0.0596, 0.0710, 0.0742, 0.0737, 0.8834, 0.9211, 0.0657, 0.8176]
8. [0.8976, 0.9251, 0.0663, 0.8531, 0.9312, 0.0614, 0.0608, 0.0744, 0.9348, 0.0303, 0.0645, 0.8045]
9. [0.8539, 0.8696, 0.9285, 0.9197, 0.0612, 0.9327, 0.9425, 0.0698, 0.9328, 0.0650, 0.0664, 0.8624]

Look at edge 10, in the first run it seems relevant, but most of the runs from 2 to 9 shows it is not. 

What I am doing is to use a Monte Carlo method to obtain a distribution, then deciding. First I was putting it outside GNNExplainer but I found it useful and now I have it integrated, which is convenient for me.

What do you think? It is a proper contribution or it should remain outside GNNExplainer?

Best Regards,
Joaquín.

**Comment by rusty1s** [comment]
You are right that `GNNExplainer` results in high variance for some predictions. Can you show me an example on how you have fixed it? I'm not sure if integrating it directly is the way to go, since it is not actually part of the `GNNExplainer` algorithm.

**Comment by joaquincabezas** [comment]
If we take the example script, I would replace lines 43 to 45 with the following:

```
node_idx = 51
explanation_size = 5
mc_runs = 50

edge_mask_agr = torch.zeros(1, edge_index.size(1)).to(device)

# Execute the GNNExplainer multiple times and track the masks > 0.5
for run in range(mc_runs):
    explainer = GNNExplainer(model, epochs=100, lr=0.05)
    node_feat_mask, edge_mask = explainer.explain_node(node_idx, x, edge_index)
    edge_mask_agr = torch.add(edge_mask_agr, edge_mask.round())

# Select only "explanation_size" edges
selected_edges = torch.topk(edge_mask_agr, explanation_size).indices.tolist()
print(selected_edges)

# Create a binary mask. 1 for the selected_edges, 0 otherwise
edge_mask.fill_(0)
edge_mask[selected_edges] = 1
```

Here I am counting the number of times each edge is ending with a value over 0.5. After all the simulation I am selecting the top 5 edges and creating a binary mask for the visualization. The reason to choose the 0.5 threshold is because after looking at the evolution of the edge values (I inserted an edge_mask_tracker into explain_node) the typical behavior is the following:

![evolution_e7_n51_cora](https://user-images.githubusercontent.com/1833086/104095717-be87a780-5298-11eb-92f2-60cb09c2021e.png)

So after a certain number of epochs (also depending on the learning rate), we can say 0.5 is a safe threshold.

PS. Regarding the behavior shown at the plot, typically there is a cut-off value (here it's 0.54) from where lines tend to go up or down, especially for nodes with a small number of edges in the k-hop neighbourhood, where the edge_size parameter is too small (I will open another discussion for parameter setting).

The standard deviation of the initial edge mask values is defined by:

`std = torch.nn.init.calculate_gain('relu') * sqrt(2.0 / (2 * N))`

For node 51 at Cora (where N=12), std ~= 0.4, which is very high (for N=100 it drops to 0.14). After the sigmoid we have values around 0.5, so this means that if the cutoff is near 0.5 the variance will be high (resulting in a bimodal distribution), and if the cutoff is far from 0.5 most simulations will provide the same value. In the following picture, edge 1 has a cutoff of 0.32 and edge 2 has a unknown cutoff as no runs hit it.

![evolution_e1+e2_n51_cora](https://user-images.githubusercontent.com/1833086/104096039-6f427680-529a-11eb-97e7-455a84bd5c1d.png)

**Comment by joaquincabezas** [comment]
Hello again,

going deeper, we can find that the variance (we can attribute it to the "split" behavior depicted in the previous message) happens more frequently when the number of edges on the 2-hop neighborhood (N from now on) is small:

![split_behavior](https://user-images.githubusercontent.com/1833086/104123943-454a8c00-534e-11eb-952e-9856d93bff29.png)

I guess there are two factors here:
1. With larger N, the spread of the random samples is smaller, thus being less probable that in this spread lies the cutoff value
2. The edge_size coefficient is the same regardless the number of edges, and for smaller N the sum is so small that it has a little effect on the loss function

The original article sets the same parameters across all simulations, and in the datasets the average N is typically bigger than 100 (for syn1 and syn2, not for syn3). But in Cora the distribution is as follows:

![distribution_cora_edges](https://user-images.githubusercontent.com/1833086/104124344-b8ed9880-5350-11eb-9fd9-a54a6ef00630.png)

Thus, it makes sense that the demo shows a lot of variance with Cora dataset. We, maybe, have to rethink about using a fixed parameter for every node. Right now I found that a proper value (proper for avoiding the split behavior) for 'edge_size' is a variable between 1/N and 2/N. For example for 2/N:

![split_behavior_variable_edge_size](https://user-images.githubusercontent.com/1833086/104127306-b7789c00-5361-11eb-9b56-8d67958fe67d.png)

Then with MonteCarlo I get the distribution and select the best K edges. I am trying with a few strategies to define "best" (higher mean, no splits, higher median, maximum hits above threshold...).

**Comment by joaquincabezas** [comment]
Hello,

I made the experiments and found that the MonteCarlo method add little stability after adjusting the edge_size coefficient, at a high cost in computation time, so I suggest using only the coefficient adjustment.

In the following image, orange crosses are the values with the default edge_size coefficient (fixed at 0.05) and blue dots are for our adjusted coefficient. This variable now depends on N, being N the number of edges in the computation graph (that is, the neighbourhood that we can reach in a certain number of hops, in this setup two hops).

![imagen](https://user-images.githubusercontent.com/1833086/108417804-bb38f180-7230-11eb-88b8-6421bef88d02.png)

Probably the use of a fixed coefficient is enough for datasets where most of the nodes have a computation graph around 100 edges, but this is not the case with Cora. In the following image we compare the two scenarios:

![imagen](https://user-images.githubusercontent.com/1833086/108418273-4b773680-7231-11eb-8f4f-c7bb2f745965.png)

If we look at how the individual components of the loss function behave in the two scenarios, it is clear that in the case of small computation graphs the optimizer does not need to minimize the component related to the number of edges:

![imagen](https://user-images.githubusercontent.com/1833086/108418811-06073900-7232-11eb-82b0-317b459dc00c.png)
 
To use this variable coefficient we need to compute N before the explanation and then set the edge_size coefficient according to it. A possible value is simply 1/N or 2/N, but I also played with a formula to adjust to the very small computation graph. 

I will share my code as an example in this thread.

**Comment by rusty1s** [comment]
Awesome work, looking forward to your code! How about we add a note on this in the documentation for `GNNExplainer`?

**Comment by joaquincabezas** [comment]
I have uploaded the code in this gist:

https://gist.github.com/joaquincabezas/aae4cd1da5653f692653165c415b775f

Please note this will not remove the instability for nodes with small neighborhoods (10-100 edges) but it will reduce it a lot, as shown in the previous plots.

I tested it against a few stability metrics to quantify 'how much' it was improving. Of course these are not formal definitions of stability but allowed me expressing the behavior I was studying.

The idea is that, as shown in the previous plots, if an edge is showing as irrelevant in some explanations and relevant in others, this is "unstable". To measure it, we execute a number of experiments and note how many times that edge is relevant, and how many times is irrelevant. We compute the P(X) of the node being relevant and then apply a function:

![image](https://user-images.githubusercontent.com/1833086/108500427-0b0acd80-72b0-11eb-9744-822e32d436a6.png)

If we do this for all the edges of a computation graph around a node and average it, and we repeat this for a subset of the graph, we will obtain a "stability metric" (between 0 and 1, being 1 perfect stability). I obtained the following results for Cora (sample of 300 nodes) and using GAT (from @PetarV-):

![image](https://user-images.githubusercontent.com/1833086/108500611-545b1d00-72b0-11eb-98ae-a79bce29efe3.png)

As expected, the Shannon entropy penalizes more than the triangular function (higher values of instability), so it show smaller values for stability.

I have seen there is another thread discussing the stability of GNNExplainer, I will announce this proposal there and see if they like it. Then we could add a note suggesting this.

**Comment by sumismyname** [comment]
Could there be more than one subgraph that explains equally well a given node in your dataset? If so, we shouldn't expect the same nodes/edges to appear in every run for different seeds, and averaging them is probably meaningless. Perhaps you can try to visualize the runs using PCA/tSNE and see if they form clusters that represent the different subgraphs.

**Comment by joaquincabezas** [comment]
GNNExplainer by itself does not search for subgraphs, so you are right, there can be a scenario where the result is a mixed set of edges from two ground truth subgraphs. While this may be a problem for studying molecules, it is not that bad for social networks.

I would suggest looking for other methods that are biased towards connected subgraphs, like SubgraphX (https://arxiv.org/abs/2102.05152) that claims "In addition, our explanations are connected subgraphs while PGExplainer and GNNExplainer identify discrete edges". Disclaimer: SubgraphX is not peer-reviewed yet and code is not available (but I personally like the approach).

Regarding the use of PCA for visualization I don't understand how you would proceed, could you please put an example?. If we are running the experiment multiple times (i.e. Montecarlo) we could simply set a color based on the percentage of times the edge is labeled as relevant.

**Comment by joaquincabezas** [comment]
Hello again @sumismyname 

SubgraphX (On Explainability of Graph Neural Networks via Subgraph Explorations) has already been peer reviewed (PMLR, ICML) and the code is publicy available! I suggest taking a look at this notebook: https://github.com/divelab/DIG/blob/dig/benchmarks/xgraph/subgraphx.ipynb

On Wednesday 22.07.2021 the authors willbe presenting the poster at ICML, so it would be a good time to ask in case you are attending the conference. Also, if you do not know DIG (https://github.com/divelab/DIG) I highly recommend it, it is being quite useful for my research.

**Comment by joaquincabezas** [comment]
> Awesome work, looking forward to your code! How about we add a note on this in the documentation for `GNNExplainer`?

Hi again! I've seen that GNNExplainer continues to be used (sometimes as baseline) in new works. So I consider it may be still worth it to add the note about the coefficients (in particular the edge_size). Let me just create a PR

**Comment by joaquincabezas** [comment]
Thanks, it is already merged!
